synapse/handlers/_base.py

```diff
@@ -282,3 +284,41 @@ def log_failure(f):
         federation_handler.handle_new_event(
             event, destinations=destinations,
         )
+
+    @defer.inlineCallbacks
+    def maybe_kick_guest_users(self, event, current_state):
+        # Technically this function invalidates current_state by changing it.
+        # Hopefully this isn't that important to the caller.
+        if event.type == EventTypes.GuestAccess:
+            guest_access = event.content.get("guest_access", "forbidden")
+            if guest_access != "can_join":
+                for member_event in current_state:
+                    try:
+                        if member_event.type != EventTypes.Member:
+                            continue
+                        if member_event.content["membership"] not in {
+                            Membership.JOIN,
+                            Membership.INVITE
+                        }:
+                            continue
+                        if (
+                            "kind" not in member_event.content
+                            or member_event.content["kind"] != "guest"
+                        ):
+                            continue
+                        message_handler = self.hs.get_handlers().message_handler
+                        event = yield message_handler.create_and_send_event(
+                            {
+                                "type": EventTypes.Member,
+                                "state_key": member_event.user_id,
+                                "content": {
+                                    "membership": Membership.LEAVE,
+                                    "kind": "guest"
+                                },
+                                "room_id": event.room_id,
+                                "sender": member_event.user_id
```

Comment:
Shouldn't the ``sender`` be the user doing the kicking rather than the user being kicked.